# HG changeset patch
# Parent 95eb747f8f84b46abc032ec28dd61b153b8d9858
support ff, enc and bin in modeline

diff --git a/src/buffer.c b/src/buffer.c
index 158c30a5a..c07ffe143 100644
--- a/src/buffer.c
+++ b/src/buffer.c
@@ -326,6 +326,29 @@ open_buffer(
 	    /* Go to the buffer that was opened. */
 	    aucmd_prepbuf(&aco, old_curbuf.br_buf);
 	    do_modelines(0);
+
+	    /* specified ff and enc, bin in modelines */
+	    if (file_ff_differs(curbuf, TRUE))
+	    {
+		/* reload buffer */
+		if (eap)
+		{
+		    /* restore ++ff and ++enc, ++bin if specified */
+		    if (eap->force_enc)
+		    {
+			char_u *fenc = enc_canonize(eap->cmd + eap->force_enc);
+			if (fenc) {
+			    vim_free(curbuf->b_p_fenc);
+			    curbuf->b_p_fenc = fenc;
+			}
+		    }
+		    if (eap->force_ff)
+			set_fileformat(eap->force_ff, OPT_LOCAL);
+		    if (eap->force_bin)
+			curbuf->b_p_bin = eap->force_bin;
+		}
+		buf_reload(curbuf, curbuf->b_orig_mode);
+	    }
 	    curbuf->b_flags &= ~(BF_CHECK_RO | BF_NEVERLOADED);
 
 #ifdef FEAT_EVAL
